### PR TITLE
feat: allow each node to control its own title

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

It would be nice to show better titles for insights in notebooks

## Changes

* fix the node id problem
* gets the notebook title when inserting and shows it

## How did you test this code?

👀 locally

### CODE REVIEW

File: frontend/src/scenes/notebooks/Nodes/NotebookNodeInsight.tsx [hunk 1]
```diff
@@ -31,11 +34,25 @@ export const NotebookNodeInsight = createPostHogWidgetNode<NotebookNodeInsightAt
     startExpanded: true,
     attributes: {
         id: {},
+        title: {},
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Nodes/NotebookNodeInsight.tsx:37]: I'd be in favour of making this a default attribute similar to what we do for `height` and `nodeId`:
https://github.com/PostHog/posthog/blob/cc674bd339e7b15b40dbf22d68aba14a66315e42/frontend/src/scenes/notebooks/Nodes/NodeWrapper.tsx#L224-L225

File: frontend/src/scenes/notebooks/Nodes/NotebookNodeInsight.tsx [hunk 2]
```diff
@@ -31,11 +34,25 @@ export const NotebookNodeInsight = createPostHogWidgetNode<NotebookNodeInsightAt
...
     },
     pasteOptions: {
         find: urls.insightView('(.+)' as InsightShortId),
         getAttributes: async (match) => {
-            return { id: match[1] as InsightShortId }
+            const shortId = match[1] as InsightShortId
```
Comment (daibhin) [inline frontend/src/scenes/notebooks/Nodes/NotebookNodeInsight.tsx:42]: If we do the above we could move this to the logic or into a `useEffect` hook on the `NodeWrapper`. Something like:

```
useEffect(() => {
  if (!node.attrs.title) {
      const title = typeof defaultTitle === "function" ? defaultTitle(attributes) : defaultTitle
      updateAttributes({ title: title })
  }
}, [])

// then in the component you can have either

title: 'Random Node with no customization'
title: (attributes) => {
 // logic to fetch title from api
}
```

**Comment by neilkakkar** [comment]
I was facing a similar issue of wanting to dynamically generate titles.

One potential thing to work around: I quite like the convention that there's always a prefix telling us what kind of node it is. Specially for when things are collapsed.

For example, if I have a feature flag, and a code example component, both for flag `xyz`, having the same title can be confusing, and doesn't tell us what it is.

So, `Feature flag: xyz` and `Feature flag code example: xyz`.

Same for insights, `Insight: `xyz``.

I.e. the default title becomes a prefix, instead of being replaced by what's in `node.attrs.title`.

I'm also not 100% clear when something should be a prop or an attribute to a node. IIUC, the existing title is a top level prop, while the "suffix" here is coming from the attribute.

**Review by daibhin — APPROVED** [review]

**Review by daibhin — APPROVED** [review]
Reviewed in person :shipit:

**Pull request merged**